### PR TITLE
fix: Drop Python 3.7 Support

### LINKED ISSUE #209: Introduce deprecation warning for python 3.7, google-api-core 1.x and protobuf 3.x
Python 3.7 is [now unsupported](https://devguide.python.org/versions/#unsupported-versions) in the python community. We should update the README to indicate that python 3.7 support will be dropped and also introduce a deprecation warning with the same information.

Similar to:
- https://github.com/googleapis/google-cloud-python/issues/11437
- https://github.com/googleapis/google-auth-library-python/pull/1371

### PR DESCRIPTION
**Pull request by holtskinner**
BEGIN_COMMIT_OVERRIDE
fix: Drop Python 3.7 Support
fix: Require google-api-core >= 2.17.1
fix: Require pandas >= 2.0.0
fix: Require proto-plus >= 1.22.3
fix: Require numpy >= 1.23.5
fix: Require pikepdf >= 8.0.0
fix: Require Pillow >= 10.0.0
END_COMMIT_OVERRIDE

Fixes #209 🦕

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -52,33 +52,28 @@
         "google.cloud.documentai_toolbox": ["templates/*.xml.j2"],
     },
     install_requires=(
-        "google-api-core >= 1.31.5, <3.0.0dev,!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0",
-        "pandas >= 1.0.0, <3.0.0",
-        "pandas >= 1.0.0, <2.0.0; python_version<'3.8'",
-        "pyarrow >= 15.0.0, <16.0.0; python_version>='3.8'",  # Required by Pandas #237
+        "google-api-core >= 2.17.1, <3.0.0dev,!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0",
+        "pandas[performance,gcp] >= 2.0.0, <3.0.0",
+        "pyarrow >= 15.0.0, <16.0.0",  # Required by Pandas #237
```
Comment (parthea) [inline setup.py:57]: Please could you update the comment `Required by Pandas #237` with a better description, or remove the comment altogether if `pyarrow` is a direct dependency of this library.

File: setup.py [hunk 2]
```diff
@@ -52,33 +52,28 @@
...
+        "numpy >= 1.23.5, < 2.0.0",
         "intervaltree >= 3.0.0",
-        "pikepdf >= 6.2.9, < 9.0.0",
-        "pikepdf >= 6.2.9, < 7.0.0; python_version<'3.8'",
+        "pikepdf >= 8.0.0, < 9.0.0",
         "immutabledict >= 2.0.0, < 4.0.0",
-        "immutabledict >= 2.0.0, < 3.0.0dev; python_version<'3.8'",
-        "Pillow >= 9.5.0, < 11.0.0",
+        "Pillow >= 10.0.0, < 11.0.0",
         "Jinja2 >= 3.1.0, <= 4.0.0",
```
Comment (parthea) [inline setup.py:70]: There are a lot of changes to the minimum requirements of the client library. Can we add each of these changes to the release notes? Edit [this comment](https://github.com/googleapis/python-documentai-toolbox/pull/262#issue-2154722320) and include the following to [override the conventional commit messages](https://github.com/googleapis/release-please?tab=readme-ov-file#how-can-i-fix-release-notes) used for the release notes.
```
BEGIN_COMMIT_OVERRIDE
fix: Drop Python 3.7 Support
fix: Require google-api-core >= 2.17.1
fix: Require pandas >= 2.0.0
fix: Require proto-plus >= 1.22.3
fix: Require numpy >= 1.23.5
fix: Require pikepdf >= 8.0.0
fix: Require Pillow >= 10.0.0
END_COMMIT_OVERRIDE
```

File: setup.py (hunk 1 above)
Comment (holtskinner) [reply setup.py:57]: pyarrow is a direct dependency of pandas.

File: testing/constraints-3.8.txt [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
-grpc-google-iam-v1
-google-cloud-bigquery
-google-cloud-documentai
-google-cloud-storage
-numpy==1.21.6
+# Pin the version to the lower bound.
+# e.g., if setup.py has "google-cloud-foo >= 1.14.0, < 2.0.0dev",
+# Then this file should have google-cloud-foo==1.14.0
+google-api-core==2.17.1
+libcst==0.2.5
```
Comment (parthea) [inline testing/constraints-3.8.txt:9]: Is `libcst` dependency needed? You can automatically generate this file using the steps in https://github.com/googleapis/synthtool/issues/1901#issuecomment-1820742964

File: setup.py (hunk 1 above)
Comment (parthea) [reply setup.py]: Could we use the `pyarrow` extra instead? `pandas[gcp,performance,pyarrow] >= 2.0.0, <3.0.0`

https://github.com/pandas-dev/pandas/blob/89b286a699b2d023b7a1ebc468abf230d84ad547/pyproject.toml#L66C1-L66C8

File: testing/constraints-3.8.txt (hunk 3 above)
Comment (holtskinner) [reply testing/constraints-3.8.txt:9]: I had issues when running that tool with this setup.py requirements

```
 install_requires=(
        "google-api-core >= 2.17.1, <3.0.0dev",
        "pandas[performance,gcp] >= 2.0.0, <3.0.0",
        "pyarrow >= 15.0.0, <16.0.0",
        "tabulate >= 0.9.0, <1.0.0",
        "proto-plus >= 1.22.3, <2.0.0dev",
        "grpc-google-iam-v1 >= 0.12.6, < 0.14.0dev",
        "google-cloud-bigquery >= 3.5.0, < 4.0.0dev",
        "google-cloud-documentai >= 2.20.0, < 3.0.0dev",
        "google-cloud-storage >= 1.31.0, < 3.0.0dev",
        "google-cloud-vision >= 2.7.0, < 4.0.0dev",
        "numpy >= 1.23.5, < 2.0.0",
        "intervaltree >= 3.0.0",
        "pikepdf >= 8.0.0, < 9.0.0",
        "immutabledict >= 2.0.0, < 5.0.0",
        "Pillow >= 10.0.0, < 11.0.0",
        "Jinja2 >= 3.1.0, <= 4.0.0",
    ),
```

I keep getting this error:
```
lower-bound-checker update --package-name google-cloud-documentai-toolbox --constraints-file /Users/holtskinner/GitHub/python-documentai-toolbox/testing/constraints-3.8.txt
Usage: lower-bound-checker update [OPTIONS]
Try 'lower-bound-checker update --help' for help.

Error: setup.py is missing explicit lower bounds for the following packages: ['Jinja2', 'google-api-core', 'google-api-core']
```

**Comment by holtskinner** [comment]
Error Message appeared in Tests before `document.py` commit was applied.

```
=================================== FAILURES ===================================
_______ test_quickstart_sample_batch_process_metadata_matching_prefixes ________

capsys = <_pytest.capture.CaptureFixture object at 0x7f3543707c50>

    def test_quickstart_sample_batch_process_metadata_matching_prefixes(
        capsys: pytest.CaptureFixture,
    ) -> None:
        batch_process_metadata = documentai.BatchProcessMetadata(
            state=documentai.BatchProcessMetadata.State.SUCCEEDED,
            individual_process_statuses=[
                documentai.BatchProcessMetadata.IndividualProcessStatus(
                    input_gcs_source="gs://test-directory/documentai/input.pdf",
                    output_gcs_destination="gs://documentai_toolbox_samples/output/matching-prefixes/1",
                ),
                documentai.BatchProcessMetadata.IndividualProcessStatus(
                    input_gcs_source="gs://test-directory/documentai/input.pdf",
                    output_gcs_destination="gs://documentai_toolbox_samples/output/matching-prefixes/11",
                ),
            ],
        )
        wrapped_document = quickstart_sample.quickstart_sample(
>           batch_process_metadata=batch_process_metadata
        )

test_quickstart_sample.py:116: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
quickstart_sample.py:80: in quickstart_sample
    metadata=batch_process_metadata
../../google/cloud/documentai_toolbox/wrappers/document.py:581: in from_batch_process_metadata
    for process in list(metadata.individual_process_statuses)
../../google/cloud/documentai_toolbox/wrappers/document.py:581: in 
    for process in list(metadata.individual_process_statuses)
../../google/cloud/documentai_toolbox/wrappers/document.py:507: in from_gcs
    shards = _get_shards(gcs_bucket_name=gcs_bucket_name, gcs_prefix=gcs_prefix)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

gcs_bucket_name = 'documentai_toolbox_samples'
gcs_prefix = 'output/matching-prefixes/1'

    def _get_shards(gcs_bucket_name: str, gcs_prefix: str) -> List[documentai.Document]:
        r"""Returns a list of `documentai.Document` shards from a Cloud Storage folder.
    
        Args:
            gcs_bucket_name (str):
                Required. The name of the gcs bucket.
    
                Format: `gs://{bucket_name}/{optional_folder}/{target_folder}/` where gcs_bucket_name=`bucket`.
            gcs_prefix (str):
                Required. The prefix of the json files in the target_folder.
    
                Format: `gs://{bucket_name}/{optional_folder}/{target_folder}/` where gcs_prefix=`{optional_folder}/{target_folder}`.
        Returns:
            List[google.cloud.documentai.Document]:
                A list of documentai.Documents.
    
        """
        file_check = re.match(constants.FILE_CHECK_REGEX, gcs_prefix)
        if file_check is not None:
            raise ValueError("gcs_prefix cannot contain file types")
    
        byte_array = gcs_utilities.get_bytes(gcs_bucket_name, gcs_prefix)
        shards = [
            documentai.Document.from_json(byte, ignore_unknown_fields=True)
            for byte in byte_array
        ]
    
        if not shards:
            raise ValueError("Incomplete Document - No JSON files found.")
    
        total_shards = len(shards)
    
        if total_shards > 1:
            shards.sort(key=lambda x: int(x.shard_info.shard_index))
    
            for shard in shards:
                if int(shard.shard_info.shard_count) != total_shards:
                    raise ValueError(
>                       f"Invalid Document - shardInfo.shardCount ({shard.shard_info.shard_count}) does not match number of shards ({total_shards})."
                    )
E                   ValueError: Invalid Document - shardInfo.shardCount (1) does not match number of shards (6).

../../google/cloud/documentai_toolbox/wrappers/document.py:134: ValueError
-------- generated xml file: /workspace/samples/snippets/sponge_log.xml --------
```

File: setup.py [hunk 4]
```diff
@@ -52,33 +52,28 @@
...
-        "immutabledict >= 2.0.0, < 4.0.0",
-        "immutabledict >= 2.0.0, < 3.0.0dev; python_version<'3.8'",
-        "Pillow >= 9.5.0, < 11.0.0",
-        "Jinja2 >= 3.1.0, <= 4.0.0",
+        "google-api-core>=2.17.1, <3.0.0dev",
+        "pandas[performance,gcp]>=2.0.0, <3.0.0",
+        "pyarrow>=15.0.0, <16.0.0",
+        "tabulate>=0.9.0, <1.0.0",
+        "proto-plus>=1.22.3, <2.0.0dev",
+        "grpc-google-iam-v1>=0.12.6, <0.14.0dev",
```
Comment (parthea) [inline setup.py:60]:
```suggestion
        "grpc-google-iam-v1>=0.12.6, <1.0.0dev",
```

**Review by parthea — APPROVED** [review]

**Pull request merged**